http/client/src/main/java/org/wildfly/security/http/client/mechanism/digest/util/DigestHttpMechanismUtil.java

```diff
@@ -0,0 +1,152 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.http.client.mechanism.digest.util;
+
+import org.wildfly.security.digest.WildFlyElytronDigestProvider;
+import org.wildfly.security.http.client.exception.ElytronHttpClientException;
+import org.wildfly.security.http.client.utils.ElytronHttpClientConstants;
+import org.wildfly.security.http.client.utils.ElytronHttpClientRequestBuilder;
+import org.wildfly.security.http.client.utils.ElytronMessages;
+import org.wildfly.security.mechanism.AuthenticationMechanismException;
+import org.wildfly.security.sasl.digest._private.DigestUtil;
+
+import static org.wildfly.security.mechanism._private.ElytronMessages.log;
+
+import static org.wildfly.security.mechanism.digest.DigestUtil.parseResponse;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.net.URI;
+import java.net.http.HttpRequest;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Elytron client for HTTP authentication
+ *
+ * @author <a href="mailto:kekumar@redhat.com">Keshav Kumar</a>
+ */
+public class DigestHttpMechanismUtil {
+
+    public static HttpRequest createDigestRequest(URI uri, String userName, String password, String authHeader, String method, String body, Map<String, String> headers) throws AuthenticationMechanismException {
+        Map<String, String> authParams = updateAuthParams(authHeader);
+        String realm = authParams.getOrDefault("realm", null);
+        String nonce = authParams.getOrDefault("nonce", null);
+        String opaque = authParams.getOrDefault("opaque", null);
+        String algorithm = authParams.getOrDefault("algorithm", "MD5");
+        String qop = authParams.getOrDefault("qop", "qop");
+        String uriPath = getUriPath(uri);
+        String cnonce = generateCNonce();
+        String nc = new String(DigestUtil.convertToHexBytesWithLeftPadding(Integer.parseInt(authParams.getOrDefault("nc", "0")), 8));
+        String resp;
+        if (qop == null) {
+            resp = computeDigestWithoutQop(uriPath, nonce, userName, password, algorithm, realm, method);
+        } else {
+            resp = computeDigestWithQop(uriPath, nonce, cnonce, nc, userName, password, algorithm, realm, qop, method);
+        }
+
+        StringBuilder requestAuthHeader = new StringBuilder();
+        requestAuthHeader.append("Digest ");
+        requestAuthHeader.append("username=\"").append(userName).append("\", ");
+        requestAuthHeader.append("realm=\"").append(realm).append("\", ");
```

Comment:
@keshav-725  If the realm is null, then the string "null" will be appended which does not sound correct. Same goes for nonce and opaque values. We should not append them if they are null